pkg/pod/status.go

```diff
@@ -140,9 +141,98 @@ func MakeTaskRunStatus(logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev
 
 	trs.TaskRunResults = removeDuplicateResults(trs.TaskRunResults)
 
+	if resolvedTaskSpec != nil {
+		// When get the results, check if the type of result is the expected one
+		if missmatchedTypes := mismatchedTypesResults(tr, trs, resolvedTaskSpec); len(missmatchedTypes) != 0 {
+			err := fmt.Errorf("missmatched Types for these results, array and object are alpha feature %v", missmatchedTypes)
+			merr = multierror.Append(merr, err)
+		}
+
+		// When get the results, for object value need to check if they have missing keys.
+		if missingKeysObjectNames := missingKeysofObjectResults(tr, trs, resolvedTaskSpec); len(missingKeysObjectNames) != 0 {
+			err := fmt.Errorf("missing keys for these results which are required in TaskResult's properties %v", missingKeysObjectNames)
+			merr = multierror.Append(merr, err)
+		}
+	}
+
 	return *trs, merr.ErrorOrNil()
 }
 
+// mismatchedTypesResults checks and returns all the mismatched types of emitted results against specified results.
+func mismatchedTypesResults(tr v1beta1.TaskRun, trs *v1beta1.TaskRunStatus, resolvedTaskSpec *v1beta1.TaskSpec) []string {
+	neededTypes := make(map[string][]string)
+	providedTypes := make(map[string][]string)
+	// collect needed keys for object results
+	if tr.Spec.TaskSpec != nil {
```

Comment:
I think it would make more sense for this function to have only one task spec as the source of truth (and same for the function below it)

Reply:
you mean merge them before calling these functions?

Reply:
If so, I merged them into one results list

Reply:
no -- I think "resolvedTaskSpec" should be used as the source of truth for what the task spec is